File: signal-layer-server/app/schemas/content.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class ContentTemplateResponse(BaseModel):
    id: str
    name: str
    description: str
    sections: list[str]
    enabled: bool
    built_in: bool
    editable: bool = False


class ContentTemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=500)
    sections: list[Literal["summary", "trend", "momentum", "volume", "structure", "risk"]] = Field(min_length=1)


class ContentTemplateUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=500)
    sections: Optional[list[Literal["summary", "trend", "momentum", "volume", "structure", "risk"]]] = Field(default=None, min_length=1)
    enabled: Optional[bool] = None


class ArticleGenerateRequest(BaseModel):
    template_id: str = "stock_technical_overview"
    symbol: str = Field(min_length=1, max_length=32)
    symbol_name: str = Field(min_length=1, max_length=100)
    market: str = Field(default="stock", min_length=1, max_length=16)
    timeframes: list[Literal["1d", "30m", "5m"]] = Field(default_factory=lambda: ["1d"])
    kline_limit: int = Field(default=300, ge=100, le=1000)


class ArticleDraftResponse(BaseModel):
    id: int
    template_id: str
    title: str
    symbol: str
    symbol_name: str
    market: str
    as_of: datetime
    timeframes: list[str]
    structured_data: dict[str, Any]
    chart_specs: list[dict[str, Any]]
    standard_markdown: str
    platform_variants: dict[str, str]
    status: str
    created_at: datetime


class ArticleStatusUpdate(BaseModel):
    status: Literal["draft", "reviewed", "rejected"]


class ArticleContentUpdate(BaseModel):
    standard_markdown: Optional[str] = Field(default=None, max_length=50000)
    platform_variants: Optional[dict[str, str]] = None

File: signal-layer/src/api/content.ts
import { api } from './client.ts'

export interface ContentTemplate {
  id: string
  name: string
  description: string
  sections: string[]
  enabled: boolean
  built_in: boolean
  editable: boolean
}

export interface ArticleFact {
  label: string
  value: string
  source: string
  timeframe: string
}

export interface ArticleSection {
  code: string
  title: string
  timeframe: string
  facts: ArticleFact[]
  paragraphs: string[]
}

export interface ArticleChartSpec {
  id: string
  title: string
  symbol: string
  symbol_name: string
  timeframe: '1d' | '30m' | '5m'
  kline_limit: number
  indicators: { type: string; params: Record<string, number> }[]
  chan_features: string[]
  aspect_ratio?: '16:9'
}

export interface ArticleDraft {
  id: number
  template_id: string
  title: string
  symbol: string
  symbol_name: string
  market: string
  as_of: string
  timeframes: string[]
  structured_data: { summary?: string; sections?: ArticleSection[]; generator?: string }
  chart_specs: ArticleChartSpec[]
  standard_markdown: string
  platform_variants: Record<'wechat' | 'zhihu' | 'xueqiu', string>
  status: 'draft' | 'reviewed' | 'rejected'
  created_at: string
}

export const fetchContentTemplates = () => api.get<ContentTemplate[]>('/content/templates')
export const createContentTemplate = (values: { name: string; description: string; sections: string[] }) => api.post<ContentTemplate>('/content/templates', values)
export const updateContentTemplate = (id: string, values: { name?: string; description?: string; sections?: string[]; enabled?: boolean }) => api.put<ContentTemplate>(`/content/templates/${id}`, values)
export const deleteContentTemplate = (id: string) => api.delete(`/content/templates/${id}`)
export const fetchArticleDrafts = () => api.get<ArticleDraft[]>('/content/drafts', { limit: '30' })
export const generateArticleDraft = (values: { template_id: string; symbol: string; symbol_name: string; market: string; timeframes: string[]; kline_limit: number }) => api.post<ArticleDraft>('/content/generate', values)
export const updateArticleStatus = (id: number, status: ArticleDraft['status']) => api.put<ArticleDraft>(`/content/drafts/${id}/status`, { status })
export const updateArticleContent = (id: number, values: { standard_markdown?: string; platform_variants?: Record<string, string> }) => api.put<ArticleDraft>(`/content/drafts/${id}`, values)
